### DISCUSSION TITLE
Add command line tool from npm package to environment

### DISCUSSION THREAD

**Discussion by svenjacobs** [pkgxdev/pkgx#793] [Help] [answered]
I'm currently evaluating `pkgx` and it looks very promising. I was wondering if it's possible to add an arbitrary command from an npm package to the current environment either through `env` or via a `pkgx.yaml` for `dev`?

So let's say I want to use [firebase-tools](https://www.npmjs.com/package/firebase-tools).
I can either call it via

```
❯ pkgx npx firebase-tools -- --version
12.6.2
```

or

```
❯ env +node
❯ npx firebase-tools -- --version
12.6.2
```

but can I also make it available in the current environment as if it was installed globally?

```
❯ firebase --version
cmd not found: firebase
```

`firebase` is the command that is provided by the `firebase-tools` package.

**Comment by jhheider** [comment]
I believe this would be: `pkgx npm i -g firebase-tools` to install the npm package into npm's global cache. Or firebase-tools could be packaged into the pantry.

**Reply by svenjacobs** [reply]
Thanks for your response. I'm preferably looking for a solution that doesn't require an additional command (every time) when entering the `dev` environment. So currently the only solution would be to provide a package for `firebase-tools` in the pantry?

**Reply by jhheider** [reply]
Or install it with npm i -g. That'll install to npms local global store.

**Reply by svenjacobs** [reply]
> Or install it with npm i -g. That'll install to npms local global store.

Yes, you're right. I didn't communicate my intentions right: I'm looking for a solution where we in a team, after every developer installed `pkgx` and integrated it into the shell, just need to enter our project dir and don't need to perform any additional initial setup steps like `npm i -g`.

Anyway, I created a [pull request](https://github.com/pkgxdev/pantry/pull/3600) for `firebase-tools` 😃

**Reply by svenjacobs** [reply, marked as answer]
At least for `firebase` there is now a solution: this is packaged as [google.com/firebase-tools](https://github.com/pkgxdev/pantry/pull/3604).

**Comment by mxcl** [comment]
This would be a neat feature for sure.

One way would be to allow `dev` to create shell functions, then:

```sh
firebase() {
    npx -- firebase "$@"
}
```

Could be defined inside that working directory. This isn’t ideal since it means the tool only works at your prompt and not deeper, but it may be enough for your needs.

Another way would be to make a shim `.bin/firebase` and add `.bin` to the `PATH` in your keyfile YAML-FM.

```yaml
env:
  PATH: {{srcroot}}/.bin:${{PATH}
```

A third (more official) way is to package firebase for the pantry, we've already done this for a few npm packages and have no issue adding such things.

I would like if npm (etc) tools could be added to the dev env automatically since this is part of the appeal of `dev` so I'll keep thinking on it.

**Reply by svenjacobs** [reply]
That sounds great! I also had the idea that maybe aliases could be defined in `pkgx.yaml` that resolve to the `pkgx` command?

```yaml
aliases:
  firebase: npx@latest firebase-tools@latest --
```

so when executing `firebase` within the environment, this resolves to `pkgx npx@latest firebase-tools@latest --`

**Reply by jhheider** [reply]
This is a cool idea. Two things we have packaged that allow convenient shipping of command runners are [xcfile.dev](https://github.com/pkgxdev/pantry/blob/main/projects/xcfile.dev/package.yml) and [just.systems](https://github.com/pkgxdev/pantry/blob/main/projects/just.systems/package.yml). So, there's lots of ways to solve this problem.
